ivadomed/models.py

```diff
@@ -769,3 +769,37 @@ def __init__(self, in_size, out_size, kernel_size=(1, 1, 1), is_batchnorm=True):
     def forward(self, inputs):
         outputs = self.conv1(inputs)
         return outputs
+
+def set_model_for_retrain(m, retrain_fraction):
+    """Set model for transfer learning.
+
+    The first layers (defined by 1-retrain_fraction) are frozen (i.e. requires_grad=False).
+    The weights of the last layers (defined by retrain_fraction) are reset.
```

Comment:
For my personal information, what is the rational to retrain only last layers, vs first, middle, randomly selected layers?

Reply:
Oh.. not a strong reason.

I implemented it this way because of the general knowledge: the first few layers learn features for general simple visual building blocks (e.g. edges, corners) vs. the last layers learn more abstract / complicated features, and more task-dependent.
So the intuition behind it is to benefit from the "simple features" learnt on the dataset1 and only fine tune the last abstract layers on the dataset2 via domain adaptation.
For instance: paragraph 2.4 of https://arxiv.org/pdf/1702.07841.pdf

Reply:
With this reasoning, in the UNet case, the deepest layers should be retrained (middle layers) and not the last layers which corresponds to the decoder layers of the less abstract features.

Reply:
As the decoder reconstructs the segmentation from the complicated / task dependent features of the bottom of the U, while receiving info from the top layers from the skip connections --> it still makes sense to me to focus on this section.
I quickly had a look, here a paper using Unet: https://arxiv.org/pdf/1906.04301.pdf see section 2.3 and Fig 1.
Heaps of paper focused on "how much of my model should I freeze?" or "how much data should I used to retrain my model". And I believe this is still an open question :/
I am happy to look deeper into the literature for the specific case of the U-net :)